File: backend/routes/locations.py
from flask import Blueprint, request, jsonify
from backend.db import get_db
from backend.auth import require_auth, require_role

locations_bp = Blueprint("locations", __name__)


@locations_bp.route("/", methods=["GET"])
@require_auth
def list_locations():
    db = get_db()
    rows = db.execute("""
        SELECT l.id, l.name, l.max_occupancy, l.active,
               COUNT(p.id) as current_occupancy
        FROM locations l
        LEFT JOIN passes p ON p.location_id = l.id AND p.status = 'active'
        WHERE l.active = 1
        GROUP BY l.id
        ORDER BY l.name
    """).fetchall()
    return jsonify([dict(r) for r in rows])


@locations_bp.route("/", methods=["POST"])
@require_role("admin")
def create_location():
    data = request.get_json(silent=True) or {}
    name = (data.get("name") or "").strip()
    if not name:
        return jsonify({"error": "Location name is required"}), 400
    try:
        max_occ = int(data.get("max_occupancy", 3))
    except (ValueError, TypeError):
        max_occ = 3
    if max_occ < 1:
        return jsonify({"error": "max_occupancy must be at least 1"}), 400

    db = get_db()
    try:
        cur = db.execute(
            "INSERT INTO locations (name, max_occupancy) VALUES (?, ?)", (name, max_occ)
        )
        db.commit()
    except Exception as exc:
        if "UNIQUE" in str(exc):
            return jsonify({"error": "A location with that name already exists"}), 409
        raise
    return jsonify({"id": cur.lastrowid, "name": name, "max_occupancy": max_occ, "active": 1}), 201


@locations_bp.route("/<int:loc_id>", methods=["PATCH"])
@require_role("admin")
def update_location(loc_id):
    db = get_db()
    row = db.execute("SELECT * FROM locations WHERE id = ?", (loc_id,)).fetchone()
    if not row:
        return jsonify({"error": "Location not found"}), 404

    data = request.get_json(silent=True) or {}
    name = data["name"].strip() if "name" in data else row["name"]
    try:
        max_occ = int(data["max_occupancy"]) if "max_occupancy" in data else row["max_occupancy"]
    except (ValueError, TypeError):
        max_occ = row["max_occupancy"]
    active = int(bool(data["active"])) if "active" in data else row["active"]

    if max_occ < 1:
        return jsonify({"error": "max_occupancy must be at least 1"}), 400

    db.execute(
        "UPDATE locations SET name = ?, max_occupancy = ?, active = ? WHERE id = ?",
        (name, max_occ, active, loc_id),
    )
    db.commit()
    return jsonify({"id": loc_id, "name": name, "max_occupancy": max_occ, "active": active})

File: src/routes/locations.js
'use strict';

const express = require('express');
const { getDb } = require('../db');
const { requireAuth, requireRole } = require('../middleware/auth');

const router = express.Router();

// All location routes require authentication
router.use(requireAuth);

// GET /api/locations - list all active locations with current occupancy
router.get('/', (req, res) => {
  const db = getDb();
  const locations = db
    .prepare(
      `SELECT l.id, l.name, l.max_occupancy, l.active,
              COUNT(p.id) as current_occupancy
       FROM locations l
       LEFT JOIN passes p ON p.location_id = l.id AND p.status = 'active'
       WHERE l.active = 1
       GROUP BY l.id
       ORDER BY l.name`
    )
    .all();
  return res.json(locations);
});

// POST /api/locations - create location (admin only)
router.post('/', requireRole('admin'), (req, res) => {
  const { name, max_occupancy } = req.body || {};
  if (!name) {
    return res.status(400).json({ error: 'Location name is required' });
  }
  const max = parseInt(max_occupancy, 10) || 3;
  if (max < 1) {
    return res.status(400).json({ error: 'max_occupancy must be at least 1' });
  }

  const db = getDb();
  try {
    const result = db
      .prepare('INSERT INTO locations (name, max_occupancy) VALUES (?, ?)')
      .run(name.trim(), max);
    return res.status(201).json({ id: result.lastInsertRowid, name: name.trim(), max_occupancy: max, active: 1 });
  } catch (err) {
    if (err.message.includes('UNIQUE')) {
      return res.status(409).json({ error: 'A location with that name already exists' });
    }
    throw err;
  }
});

// PATCH /api/locations/:id - update location (admin only)
router.patch('/:id', requireRole('admin'), (req, res) => {
  const db = getDb();
  const location = db.prepare('SELECT * FROM locations WHERE id = ?').get(req.params.id);
  if (!location) {
    return res.status(404).json({ error: 'Location not found' });
  }

  const { name, max_occupancy, active } = req.body || {};
  const newName = name !== undefined ? name.trim() : location.name;
  const newMax = max_occupancy !== undefined ? parseInt(max_occupancy, 10) : location.max_occupancy;
  const newActive = active !== undefined ? (active ? 1 : 0) : location.active;

  if (newMax < 1) {
    return res.status(400).json({ error: 'max_occupancy must be at least 1' });
  }

  db.prepare('UPDATE locations SET name = ?, max_occupancy = ?, active = ? WHERE id = ?').run(
    newName,
    newMax,
    newActive,
    req.params.id
  );

  return res.json({ id: location.id, name: newName, max_occupancy: newMax, active: newActive });
});

module.exports = router;
